### PR TITLE
Tzc update

### PR DESCRIPTION
**Pull request by MrVan**
1. Do not touch reserved bits for region 0.
2. Export more API for flexible configuration.

### CODE REVIEW

File: core/drivers/tzc380.c [hunk 1]
```diff
@@ -117,8 +117,13 @@ void tzc_configure_region(uint8_t region, vaddr_t region_base, uint32_t attr)
 
 	assert(region < tzc.num_regions);
 
-	tzc_write_region_base_low(tzc.base, region, addr_low(region_base));
-	tzc_write_region_base_high(tzc.base, region, addr_high(region_base));
+	if (!region) {
```
Comment (jenswi-linaro) [inline core/drivers/tzc380.c]: It wouldn't hurt with a comment here with something about region0

File: core/drivers/tzc380.c [hunk 2]
```diff
@@ -92,6 +99,38 @@ void tzc_init(vaddr_t base)
 			   BUILD_CONFIG_NR_MASK) + 1;
 }
 
+void tzc_security_inversion_en(vaddr_t base)
+{
+	write32(1, base + SECURITY_INV_EN_OFF);
+}
+
+void tzc_enable_region(uint8_t region)
```
Comment (jenswi-linaro) [inline core/drivers/tzc380.c]: `tzc_region_enable()` would be more consistent with for instance `tzc_security_inversion_en()` and `tzc_int_clear()`

Comment (jenswi-linaro) [reply core/drivers/tzc380.c]: It wouldn't hurt with a comment with a short description of some of the more important functions

**Comment by MrVan** [comment]
Updated with more comments added. And tzc_enable_region -> tzc_region_enable

**Comment by jenswi-linaro** [comment]
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: core/drivers/tzc380.c [hunk 3]
```diff
@@ -59,6 +61,11 @@ static void tzc_write_action(vaddr_t base, enum tzc_action action)
 	write32(action, base + ACTION_OFF);
 }
 
+static uint32_t tzc_read_region_attributes(vaddr_t base, uint32_t region)
```
Comment (etienne-lms) [inline core/drivers/tzc380.c]: minor: maybe move right above `tzc_write_region_attributes`

File: core/drivers/tzc380.c [hunk 4]
```diff
@@ -92,6 +99,55 @@ void tzc_init(vaddr_t base)
...
+ */
+void tzc_fail_dump(void)
+{
+	vaddr_t base __maybe_unused = core_mmu_get_va(tzc.base,
+						      MEM_AREA_IO_SEC);
+
+	EMSG("Fail address Low %x\n", read32(base + 0x20));
+	EMSG("Fail address High %x\n", read32(base + 0x24));
+	EMSG("Fail Control %x\n", read32(base + 0x28));
+	EMSG("Fail Id %x\n", read32(base + 0x2C));
```
Comment (etienne-lms) [inline core/drivers/tzc380.c]: Could use define macros for the offsets?
```
#define TZC_FAIL_ADDRESS_LOW	0x20
#define TZC_FAIL_ADDRESS_HIGH	0x24
#define TZC_FAIL_CONTROL		0x28
#define TZC_FAIL_ID				0x2C
```

File: core/drivers/tzc380.c [hunk 5]
```diff
@@ -117,9 +173,20 @@ void tzc_configure_region(uint8_t region, vaddr_t region_base, uint32_t attr)
...
 	assert(region < tzc.num_regions);
 
-	tzc_write_region_base_low(tzc.base, region, addr_low(region_base));
-	tzc_write_region_base_high(tzc.base, region, addr_high(region_base));
-	tzc_write_region_attributes(tzc.base, region, attr);
+	/*
+	 * For region 0, this high/low/size/en field is Read Only (RO).
+	 * So should not configure those field for region 0.
+	 */
+	if (!region) {
```
Comment (etienne-lms) [inline core/drivers/tzc380.c]: s/`!region`/`region`/

**Comment by MrVan** [comment]
Updated. Use macros for tzc_fail_dump. "s/!region/region/", in my local, I use "region != 0", my bad that not use up to date patch, in last pull request.

**Comment by jforissier** [comment]
@etienne-lms ping

File: core/drivers/tzc380.c [hunk 6]
```diff
@@ -92,6 +99,55 @@ void tzc_init(vaddr_t base)
...
+/*
+ * Dump info when TZC380 catchs an unallowed access with TZC
+ * interrupt enabled.
+ */
+void tzc_fail_dump(void)
+{
+	vaddr_t base __maybe_unused = core_mmu_get_va(tzc.base,
+						      MEM_AREA_IO_SEC);
+
+	EMSG("Fail address Low %x\n", read32(base + FAIL_ADDRESS_LOW_OFF));
```
Comment (etienne-lms) [inline core/drivers/tzc380.c]: Replace `%x` with `%" PRIx32 "`, here and below.

**Comment by MrVan** [comment]
Updated with "Replace %x with 0x%" PRIx32 ", 2nd commit log refined.

**Comment by jforissier** [comment]
Thanks @MrVan . I think we can consider @etienne-lms's "looks ok to me" as an `Acked-by: Etienne Carriere <etienne.carriere@linaro.org>`. Can you please add it?

**Comment by MrVan** [comment]
@jforissier Updated. Thanks.

**Pull request merged**